awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -640,47 +316,22 @@ def create_or_update_if_needed(self, existing_item, new_item, endpoint=None, ite
         else:
             return self.create_if_needed(existing_item, new_item, endpoint, on_create=on_create, item_type=item_type, associations=associations)
 
-    def logout(self):
-        if self.oauth_token_id is not None and self.username and self.password:
-            # Attempt to delete our current token from /api/v2/tokens/
-            # Post to the tokens endpoint with baisc auth to try and get a token
-            api_token_url = (
-                self.url._replace(
-                    path='/api/v2/tokens/{0}/'.format(self.oauth_token_id),
-                    query=None  # in error cases, fail_json exists before exception handling
-                )
-            ).geturl()
-
-            try:
-                self.session.open(
-                    'DELETE',
-                    api_token_url,
-                    validate_certs=self.verify_ssl,
-                    follow_redirects=True,
-                    force_basic_auth=True,
-                    url_username=self.username,
-                    url_password=self.password
-                )
-                self.oauth_token_id = None
-                self.authenticated = False
-            except HTTPError as he:
-                try:
-                    resp = he.read()
-                except Exception as e:
-                    resp = 'unknown {0}'.format(e)
-                self.warn('Failed to release tower token: {0}, response: {1}'.format(he, resp))
-            except(Exception) as e:
-                # Sanity check: Did the server send back some kind of internal error?
-                self.warn('Failed to release tower token {0}: {1}'.format(self.oauth_token_id, e))
+    def exit_tower_exception(self, tower_exception):
+        response_data = {}
+        if tower_exception.response:
+            response_data['response'] = tower_exception.response
```

Comment:
I thought this was a really good idea, and I want to get something like it implemented.

However, I looked through the code, and it looks like attaching a `response` attribute to the exception object was never hooked up right, unless I'm reading something wrong.

I think the response and status code would be great to include in this data.

Reply:
Updated a lot of references to include the response. Initially it was only returned in some of the last two exceptions in make_request. Now all of the non-200 error codes will also return this.